src/main/java/com/ververica/flink/table/gateway/deployment/ClusterDescriptorAdapterFactory.java

```diff
@@ -0,0 +1,45 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.ververica.flink.table.gateway.deployment;
+
+import org.apache.flink.configuration.Configuration;
+import org.apache.flink.configuration.DeploymentOptions;
+
+/**
+ * The factory to create {@link ClusterDescriptorAdapter} based on execution.target.
+ */
+public class ClusterDescriptorAdapterFactory {
+    public static ClusterDescriptorAdapter createClusterDescriptorAdapter(Configuration configuration){
```

Comment:
I think each job operation should hold a `ClusterDescriptorAdapter` instance which should hold an immutable ClusterID. For session cluster, we can give the global cluster id from `ExecutionContext`, for yarn-per-job cluster, we can give a future when creating, and complete the future when the job is submitted. `ClusterDescriptorAdapter` has getJobStatus, cancelJob method, and each kind of  `ClusterDescriptorAdapter` implement different logic. After that, `InsertOperation#containsYarnApplicationNotFoundException` and `InsertOperation#isYarnPerJobMode` can also be moved to `YarnPerJobClusterDescriptorAdapter`

Reply:
Thanks for your very good advice. I will continue to improve these codes.